Monitoring Custom Metrics

waprin opened the pull request:
This kicks off the PR for #1867 with the usage docs. Implementation is started and will follow this soon.

@rimey is primary reviewer here, largely just took most of his suggestions but some followup comments here:

1) Keyword arguments sounds like a good idea and not too tricky, we can do that now.

2) re: a way to write multiple TimeSeries in one go. I did add the suggested method.

Currently, there is no way to write more than one point in a TimeSeries. The only way we could batch writes is doing something like this: 

https://developers.google.com/api-client-library/python/guide/batch

This might involve some changes to the core client library here, how important do we think this is?

In general, what are the main use cases for writing multiple TimeSeries at once? Is it just batching to reduce network requests? I think the method would currently just be a wrapper around

```
for ts in timeseries_list:
     ts.create()
```

which may be a good wrapper to have, even though users could trivially write that out.

3) I think you want the most common use case to be as simple as possible, which is why I think we want a `write` method that just takes a single value without forcing to pass the metric and resource every time. Since the `timeseries.create()` API call requires a Metric and Resource, we have to have a class that contains that Metric and Resource. We can of course just use the TimeSeries itself, but since you can only write a single point to that TimeSeries, it's not something that makes sense to re-use if we want the local object to map well to the API object. So I instead propose to add a TimeseriesWriter class that can take a Metric and Resource and be re-used to create new TimeSeries with those Metric and Resource. 

```
timeseries_writer = client.get_timeseries_writer(metric, resource)
timeseries_writer.write_point(value) # API call
````

cc @sharbison3 @supriyagarg

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,112 @@ follows::
+                                 value_type=ValueType.INT64,
+                                 labels=[label],
+                                 description='Cumulative count of HTTP responses.')
+>>> if not metric_descriptor.exist():
```
exist -> exists

**Note**: There is no such method for the `MetricDescriptor` class currently.

waprin replied:
Yes, exist was typo thanks.

I have added the method.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,112 @@ follows::
+You can also use the :class:`~gcloud.monitoring.TimeseriesWriter` class to
+associate a fully specified Metric and Resource pair to write :class:`~gcloud.monitoring.timeseries.TimeSeries` together::
+
+    >>> timeseries_writer = client.get_timeseries_writer(metric, resource)
```
get_timeseries_writer -> timeseries_writer 

for consistency with methods like `client.group`, and `client.metric_descriptor`.

waprin replied:
done (acking in comments but will do a commit amend and force push at end)

waprin commented:
I realized an issue. Metric currently doesn't contain the descriptor, which means when you try to write a value for it, it doesn't know how to specify its value type. So I have changed Metric to now optionally include a MetricDescriptor. It's still a namedtuple but can have optional default fields (also made this change to TimeSeries since it has output only field).

Not sure if there's opposition to the Metric object keeping a reference to its MetricDescriptor, but it greatly simplifies everything for the Metric to be able to hold onto that information, will update the usage docs accordingly.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,112 @@ follows::
+    >>> value = get_num_404s_in_last_five_minutes() # get value from a hypothetical call
+    >>>
+    >>> client.write_timeseries(value, start_time=start, end_time=end,
+    ...                         metric=metric, resource=resource) # API call
```
For writing a single point, it might be better to just go with the TimeSeries object as you do below (L377-378). Otherwise `timeseries_writer` and `write_timeseries` can become confusing.

Also, the other APIs that have implemented creation (Group & MetricDescriptor) don't support creation in the client class.

waprin replied:
Ok, I was trying to get some opinions on it. The motivation for going in this direction was because it seems you have to consistently pass in Metric and Resource into every API call.  Seems if the common use case was to write points for a metric over time, and if those keep taking the same arguments every call, it would be better to have something to encapsulate those arguments. 

I agree it's a bit inconsistent, but it reflects the TimeSeries being slightly inconsistent at an API level- it's one of the few API calls where you call `create` multiple times to update it (most API objects are created once with a unique ID and after that only updated). TimeSeries should really have a write_point method of some sort, but it doesn't, so to do multiple writes to the same TimeSeries we can either copy the TimeSeries to a new object but replace the points (similar to what the `header` method does now but will need to add keyword argument magic), or keep recreating similar objects with the same arguments. Both seemed clumsy to me so I was putting an alternative out there but sounds like you're not a fan. I will mull it over and curious if Ken has an opinion, ultimately not a big deal either way and fine to get rid of it if you feel strongly.

waprin replied:
I said TimeSeries should have a `write_point` method at the API level, but I guess it's also possible we could just add one at the object level that does a copy of itself with the new points and calls `create` on the new copy. Something like:

```
    >>> ts1 = client.timeseries(value, end_time=end_time, metric=metric1, resource=resource1)
   >>> ts1.write_point(value2) # API call, creates new TS with same metric/resource
```

Any better?

supriyagarg replied:
I was just talking about dropping the method `write_timeseries` for writing a single point. That can already be done using:

    ts1 = client.timeseries(value, end_time=end_time, metric=metric1, resource=resource1)
    ts1.create()

For writing points to the same timeseries, the `TimeseriesWriter` class makes sense. Adding write_point to `TimeSeries` is strange.

waprin replied:
Ah ok, now I understand, sounds good to me.

I had made a mistake and where I wrote `write_timeseries` I meant, `write_point`. But I think your point would still stand that if we're introducing a TimeSeries writer, we should just keep using it in all cases.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,112 @@ follows::
+>>> client = monitoring.Client()
+>>> label = LabelDescriptor('response_code', LabelValueType.INT64,
+    ...                         description='HTTP status code')
+>>> metric_descriptor = = client.metric_descriptor(
```
There is an extra equal sign.

Also, the formatting is off in the lines below, it should just be a 4 space indent from L303.

waprin replied:
done (next commit amend)

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,112 @@ follows::
+    >>> client.write_timeseries(value, start_time=start, end_time=end,
+    ...                         metric=metric, resource=resource) # API call
+
+You can also use the :class:`~gcloud.monitoring.TimeseriesWriter` class to
```
For consistency with `TimeSeries`:

Time**s**eriesWriter -> Time**S**eriesWriter?

waprin replied:
done (next commit amend)

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,112 @@ follows::
+
+    >>> ts1 = client.timeseries(value, end_time=end_time, metric=metric1, resource=resource1)
+    >>> ts2 = client.timeseries(value2, end_time=end_time2, metric=metric2, resource=resource2)
+    >>> client.write_time_series([ts1, ts2])
```
I was going to say  the name is very similar to `write_timeseries` above, but see that the above method was to be `write_point`.

In any case, it is better to drop this method from the client, and perhaps add it to `TimeSeriesWriter`? We could even not support this for now, since the common use case is to set up a single metric stream, and write points to it.

waprin replied:
I mostly agree, I included it because @rimey suggested it but if you see in the PR notes I was a bit confused as to what he was getting at.

supriyagarg replied:
I went back and looked at the PR notes. Since there was no TimeSeriesWriter class, that designed looked good, but having:

`client.write_timeseries()` and
`client.timeseries_writer()` 

just seems very messy/confusing to me.

To keep both (and to shorten the name), how about rename:
* `TimeSeriesWriter` -> `MetricStream`, and 
* `client.timeseries_writer` -> `client.metric_stream`

`MetricStream` can have the `write_point` method.
`Client` can have the `write_timeseries` method.

waprin replied:
Sounds good to me. I am still not totally sure if we really need the write_timeseries method at all though.

waprin commented:
Though I'm realizing that if we introduce a background thread, I'll have the same problem as I did in the Logging handler, which is that we need to be able to safely copy the Client, in a way that works for both HTTP and gRPC...hmm.

supriyagarg commented:
Just to clarify: 1 point/timeseries/second means the rpc call be at most once per second, right?

In such a case, TimeSeriesWriter should be able to handle it.
However, TimeSeries.create() will not have enough context to do so - we will have to clearly document that the user needs to handle the delay while using it.

waprin commented:
If we wanted, `TimeSeries` objects can have the `Client` object attached to it in the factory so they could have the context if we wanted to, or we could make the `create` call take the client like `create(client)`. But I overall agree that code maybe doesn't belong there.

I'm not even totally sure if it makes sense to try to handle the rate-limiting in this library, raises several issues, kicked off a discussion #2059.

rimey commented:
First of all, I'd like to set out what I believe to be an illustrative and realistic usage scenario:

- The application has a number of replicated instances.
- Each instance sends custom metrics to the service once a minute. It only sends one point per time series.
- Each metric is identified by several labels. These might, for example, represent different categories of requests being processed, and different response statuses. There might be tens or hundreds of label combinations. Let's just assume that each instance writes them all each time. It writes them in just one request (or a few, if there are more than 200) because that's quicker and uses less request quota. That also enables the service to operate more efficiently.
- The monitored resource (the type and all the labels) is the same for all of those time series, for a given instance. The type might be `gce_instance`, `gke_container`, or `aws_ec2_instance`, as the case may be. (If the application were not replicated, one might use the `global` monitored resource type instead.)

Some thoughts:

- Batching of multiple time series in a single write is important.
- Batching of multiple points for the same time series is not important. (The API might allow that someday, for special situations where it is useful.)
- It would indeed make sense to allow the application to specify the monitored resource just once at startup. That said, I don't think it's worth introducing a class just for this reason.
- It does not make sense to try to factor out the metric in the same way, at least not in a simple client library that wraps the API straightforwardly.

I would reiterate the proposal I made in [#1867](https://github.com/GoogleCloudPlatform/gcloud-python/issues/1867#issuecomment-229784134):

```
# Write a point to a time series.
client.write_point(metric=metric, resource=resource, value=3.14) # API call

# Write data to several time series. The argument is a sequence of TimeSeries objects.
client.write_time_series([ts1, ts2, ..., tsN]) # API call
```

I don't see value in messing with metric descriptors when writing custom metric data. We can just deduce the value type from the Python type of the value. That will mean that one can't provide an `int` when the value type is `DOUBLE`, or a `float` when the value type is `INT64`, but so be it.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
     https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries
+
+
+Writing Custom Time Series
```
Writing Custom Metrics

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+Writing Custom Time Series
+---------------------------
+
+The Stackdriver Monitoring API can be used to create and delete custom metrics, as well as write
```
It can be used to create and delete _metric descriptors_ for custom metrics, but that's already discussed earlier in the document and doesn't need to be repeated here.

waprin replied:
done with slight tweak

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+...    'response_code': 404
+... })
+
+It's also possible to assign labels to Metrics and Resources with keyword arguments::
```
This enhancement of the `Metric` and `Resource` classes can be a separate pull request.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+
+    >>> start = datetime.utcnow()
+    >>> end = datetime.utcnow() - datetime.timedelta(minutes=5)
+    >>> value = get_num_404s_in_last_five_minutes() # get value from a hypothetical call
```
This is a cumulative metric, so you'd probably want to capture the start time when the application instance starts up and send values representing the cumulative total at the end time.

waprin replied:
This makes sense, but feel like we can't represent this naturally without a separate class or separate method. Don't we need something to say "now start a new start time" and a way to express "here's a new point for the start time I signified earlier"?

rimey replied:
Representing this "naturally" results in a thick client library far outside the scope of this first step.

waprin replied:
Cool, just verifying you were looking for this in a followup PR. Will put together the two simple methods you originally asked for!

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+    >>> metric_stream = client.metric_stream(metric, resource)
+    >>> metric_stream.write_point(value, start_time=start, end_time=end) # API call
+
+Stackdriver Monitoring supports several **value types**. A *CUMULATIVE* metric represents a value
```
`CUMULATIVE` is a metric kind, not a value type.

Generally, we should try hard to minimize the amount of text here that duplicates the API documentation. It's too difficult to get right, and it's going to be a headache to keep in sync.

waprin replied:
Done by killing some of the text and linking to more docs.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+
+    >>> metric_stream.write_point(value, start_time=start_time) # API call
+
+As an alternative to :class:`~gcloud.monitoring.timeseries.MetricStream`,  it's also possible
```
I doubt this would be helpful.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+ :class:`gcloud.monitoring.timeseries.TimeSeries` object, which can then written later::
+
+    >>> ts1 = client.timeseries(value, end_time=end_time, metric=metric1, resource=resource1)
+    >>> ts1.create()
```
I doubt this is needed.

waprin replied:
I took it out of the usage docs, but we likely want to have it as a public API method regardless to keep consistency with the rest of the library.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,110 @@ follows::
+ to have the :class:`gcloud.monitoring.client.Client`  return a
+ :class:`gcloud.monitoring.timeseries.TimeSeries` object, which can then written later::
+
+    >>> ts1 = client.timeseries(value, end_time=end_time, metric=metric1, resource=resource1)
```
Providing factories for Metric, Resource, and TimeSeries could be reasonable, but it could as well be a separate pull request.

waprin replied:
I kept them on this pass, if you really feel strongly let me know, but the reasons to keep them are

* Consistency across the library
* Having factory methods lets us keep the argument defaults in there while keeping the namedtuple classes cleaner.

rimey commented:
I honestly think it makes sense to get started on this without introducing any new classes at all.

waprin commented:
Everything sounds ok but:

> Batching of multiple time series in a single write is important.

I think there's no way to do this at the API level, but you can send multiple API calls in one request, but I *think* it would require changes to the core client code to do what google-api-python-client does here:

https://developers.google.com/api-client-library/python/guide/batch

also sounds like it should be a separate PR.

Sounds like the best course of action is to do something super minimal here, get it merged, figure out the more complicated stuff after that.

supriyagarg commented:
The Monitoring API's `timeSeries.create` method allows writing a single point to multiple timeseries:

https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.timeSeries/create

waprin commented:
Ah thank you, now I understand.

waprin commented:
@rimey @supriyagarg ready for another pass at your convenience. As discussed, please offer further comments, or LGTM if happy and then I will schedule in-person chat.

High level notes:

1. Killed most of what Ken asked me to. I kept the factory methods for reasons noted in comments (API consistency and argument defaulting). 

2. Kept some discussion of metric kinds in there, but if you still feel its too redundant with the docs we can remove more. Same goes for the setup code which is slightly redundant with the above code on creating MetricDescriptors.

3. One other slight difference from Ken's proposed metric is that `metric` and `resource` are positional arguments rather than keyword arguments. The reason for this is there don't seem to be any reasonable defaults - if they default to None, then we just need to raise an Exception if they're not specified, which seems like pointless code when we could just make them positional arguments. 

Still happy to change any of it if you feel strongly, let me know.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+To write multiple TimeSeries points in a single batch, you can use
+:meth:`~gcloud.monitoring.client.write_time_series`::
+
+    >>> ts1 = client.timeseries(value, metric1, reosource1, end_time=end_time)
```
reosource1 -> resource1

supriyagarg commented:
> `3.` One other slight difference from Ken's proposed metric is that metric and resource are positional arguments rather than keyword arguments. 

`metric`, `resource` and `value` are definitely required arguments - in the examples, it is still useful to include them as keyword arguments, so that there is lower chance for the users to get the order wrong.

rimey commented:
- I don't think it's helpful to capture metric descriptors in `Metric` instances. It's extra work for the user, it's another thing for them to get wrong, utilizing it would duplicate server logic (which is subject to change) on the client side, and it's just not needed.
- If you really want to make use of metric, resource, and time series factories in this usage doc (which is reasonable), I suggest sending a pull request to add them to `client.py`. This can be separate from the pull request for the code for writing custom metrics. In fact, we might even merge it before this usage doc pull request is merged.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+
+   >>> client.write_point(value, metric, resource) # API call
+
+For *DELTA* metrics, the points reflect a change during a finite time interval, so both start_time
```
Custom delta metrics are not supported.

waprin commented:
@rimey I have the PR you suggested ready, it's blocked by #2066 which broke this whole repo :-\ Agreed I don't need the descriptor in `Metric`, it was accidentally left in from when I was using it to get the value type. 

@supriyagarg thanks, now I see people just mean use keyword arguments for positional args for clarity in usage docs, obvious in hindsight! 

Review changes have been made to this doc, as soon as the above issue is fixed I will throw up the new PR as soon as #2067 gets merged.

One note, I also add the `exists` method to MetricDescriptor in the new PR. It seems useful to me for code that creates MetricDescriptor but isn't sure if it already exists. I'm not sure if we need it, especially if we suggest users[ rely on auto-creation of descriptors](https://cloud.google.com/monitoring/custom-metrics/creating-metrics#auto-creation), which I was also curious if we should be recommending anyway.

rimey commented:
These PRs shouldn't be touching upon creation or probing of metric descriptors in any way. The basic client library methods that we are discussing here (not ruling out additional methods that do more) should generally each wrap one just one API method. The usage doc can simply assume that the metric descriptor either already exists or will be auto-created.

We should specifically avoid casually suggesting any sort of strategy for explicit dynamic creation of metric descriptors. It's tricky and subtle, and just not relevant here.

waprin commented:
No auto-creation seems fine. 

I'll take `exists` out, maybe you don't want it or want it in a separate PR but let me briefly justify why I think we'll want it:

* It's a common method call that is consistent with the rest of the library, both `gcloud.pubsub.subscription` or `gcloud.storage.bucket` have this method
* The code to do it is simple, but requires a try/except which is slightly awkward to expose
* It's convenient if sample code can be run more than once by a user without breaking things, and without having to cleanup afterwards so that users can see what they created in the console/API.

tseaver commented:
@waprin I've merged #2065 to master.

waprin commented:
@supriyagarg @rimey this is ready for another pass. I moved `value` after `metric` and `resource` to align with the `time_series` factory. This might have not been necessary since `points` will likely be a new method rather than a different keyword arg but wasn't sure. Changed the MetricDescriptor links just so I could link to the title and cut out all the redundant info, and tried to work in more of the info discussed in the factory review.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+---------------------------
+
+The Stackdriver Monitoring API can be used to write data points to custom metrics. Please refer to
+the documentation for the `Creating Custom Metrics`_ for more information.
```
for the → for

rimey replied:
Better to refer to the main landing page for [Using Custom Metrics](https://cloud.google.com/monitoring/custom-metrics/).

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+custom :class:`~gcloud.monitoring.metric.MetricDescriptor` object, and then use its
+type to create a fully parameterized :class:`~gcloud.monitoring.metric.Metric` by providing
+specific values for the available labels. For more information on creating metric descriptors,
+see the `Metric Descriptors`_ section.
```
There's confusion here regarding metric descriptor objects in the client library versus metric descriptors as created through the API. I suggest the following:

To write a data point to a custom metric, you will have to provide an instance of `Metric` specifying the metric type as well as values for the metric labels. You will either have created the metric descriptor earlier or be planning to take advantage of [auto-creation](https://cloud.google.com/monitoring/custom-metrics/creating-metrics#auto-creation).

supriyagarg replied:
This comment hasn't been addressed yet, and will definitely make the text better.

waprin replied:
I attempt to address it, so if you have any other improvements in mind will need you to be more specific. I will note that this comment is on the outdated commit.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+of the available ``ResourceDescriptors`` which can be listed using the
+:class:`~gcloud.monitoring.resource.ResourceDescriptor` class, which similarly requires that its
+available labels have values specified. The ``project-id`` label is a special case that should
+not be specified.
```
I would suggest the following:

You will also need to provide a `Resource` instance specifying a monitored resource type as well as values for all of the monitored resource labels, except for `project_id`, which is ignored when writing. A good choice is to use the underlying physical resource where your application code runs – e.g., a monitored resource type of `gce_instance` or `aws_ec2_instance`. In some limited circumstances, such as when only a single process writes to the custom metric, you may choose to use the `global` monitored resource type.

supriyagarg replied:
Same here - comment not addressed. Resources are simpler to write about since the user does not have to define the resource descriptor, but the rewording explains how to pick the resource/resource type to the user.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+not be specified.
+
+>>> from gcloud import monitoring
+>>> # Create a Resource by its resource name, assign labels as dict
```
`# Create a Resource object for the desired monitored resource type.`

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+...     'instance_id': '1234567890123456789',
+...     'zone': 'us-central1-f'
+... })
+>>> # Create a Metric using its metric descriptor. assigning label as dict
```
`# Create a Metric object, specifying the metric type as well as values for any metric labels.`

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+can be used to write :class:`~gcloud.monitoring.timeseries.Point` values.
+
+When writing points, the Python type of the value must match the *value_type* specified
+in the associated :class:`~gcloud.monitoring.metric.MetricDescriptor`. For example, a Python
```
Since code for writing custom metrics might not need to deal with `MetricDescriptor` objects, I think it would be better to link to the API documentation:

...must match the value type specified in the appropriate [metric descriptor](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors). For example, ...

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+float will map to ``ValueType.DOUBLE``.
+
+Stackdriver Monitoring supports several **metric kinds**: *GAUGE*, *CUMULATIVE*, and *DELTA*.
+However, *DELTA* metrics can not be created as custom metrics.
```
"metric kinds" should be in italics, not bold. "can not" should be "cannot". Suggestion:

Stackdriver Monitoring defines several _metric kinds_: `GAUGE`, `CUMULATIVE`, and `DELTA`.
However, `DELTA` custom metrics are not supported.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+    >>> client.write_point(metric=metric, resource=resource, 3.14, end_time=end) # API call
+
+By default, end_time defaults to :meth:`~datetime.datetime.utcnow()`, so metrics written for the
+current time can be simplified to::
```
"end_time" is code. Suggestion:

...so metrics can be written for the current time as follows:

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+   >>> client.write_point(metric, resource, 3.14) # API call
+
+For *CUMULATIVE* metrics, the same start time should be re-used repeatedly as more points are
+written to the time series. In this context, the start time is sometimes referred to as the *RESET*
```
_RESET_ → _reset_

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+For *CUMULATIVE* metrics, the same start time should be re-used repeatedly as more points are
+written to the time series. In this context, the start time is sometimes referred to as the *RESET*
+time. Once the cumulative metric is ready to be reset, the *RESET* time should be reset by
+specifying a new ``start_time``. Please refer to the `TimeSeries`_ documentation for more
```
Suggestion:

...is also called the _reset_ time and is typically determined when the application process starts up. Cumulative metrics enable the monitoring system to compute rates of increase across process restarts, which would otherwise show up as huge negative spikes.

waprin replied:
worked this in by slight shuffling, take a look.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+    >>> client.write_point(metric, resource, 3, start_time=RESET, end_time=TIME1) # API call
+    >>> client.write_point(metric, resource, 6, start_time=RESET, end_time=TIME2) # API call
+    >>> # Now reset the time series, for example on an application restart.
+    >>> RESET = datetime.utcnow()
```
There's no need to have an example showing an application restart. Please instead simplify this example to first capture a reset time and then send a couple of points using it. Using the default end time seems reasonable.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+:meth:`~gcloud.monitoring.client.write_time_series`::
+
+    >>> ts1 = client.time_series(metric1, resource1, 3.14, end_time=end_time)
+    >>> ts2 = client.time_series(metric2, resource2, 42, end_time=end_time2)
```
Both time series objects should use the same monitored resource and end time. That would be typical usage.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+For more information on creating ``TimeSeries``, please refer to the API reference docs on
+`TimeSeries`_.
+
+All timezone-naive Python ``datetime`` dates are assumed to be UTC.
```
dates -> objects

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,86 @@ follows::
+the API must only include a single point.
+
+For more information on creating ``TimeSeries``, please refer to the API reference docs on
+`TimeSeries`_.
```
This reference seem unhelpful. These sorts of general references to the API documentation add verbiage. We should keep them to a minimum.

waprin replied:
Got rid of this one. But overall seems pretty harmless to link here and there /shruggie

rimey commented:
@waprin I added some comments. PTAL.

waprin commented:
@rimey all review comments addressed

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+You must also create a fully parameterized :class:`~gcloud.monitoring.resource.Resource` from one
+of the available ``ResourceDescriptors`` which can be listed using the
+:class:`~gcloud.monitoring.resource.ResourceDescriptor` class, which similarly requires that its
+available labels have values specified. The ``project-id`` label is a special case that should
```
project-id -> project_id

waprin replied:
done

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+The Stackdriver Monitoring API can be used to write data points to custom metrics. Please refer to
+the documentation on `Custom Metrics`_ for more information.
+
+To write a data point to a custom metric, you must provide an instances of
```
instances -> instance

waprin replied:
done

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+be created using the :class:`~gcloud.monitoring.metric.MetricDescriptor` class or
+auto-created.  For more information on creating metric descriptors, see the `Metric Descriptors`_
+section. For information on metric type auto-creation, see `Auto-creation of custom metrics`_.
+
```
(Responding here so I could point to the latest version.)

I see. I didn't see a response to it, so thought you might have missed it. The confusion regarding metric descriptor objects in the client library vs. metric descriptors as created through the API still remains. When talking about type / descriptor you can replace the last 3 sentences with something like:


You will either have created the metric descriptor earlier or be planning to take advantage of [auto-creation](https://cloud.google.com/monitoring/custom-metrics/creating-metrics#auto-creation).

waprin replied:
Hmm, I don't see how the confusion remains. Your suggested replacement sentences seem very similar to what is written minus the link to the class and previous reference docs. If I say "You must have created the metric descriptor earlier" it raises the question, how do I create it, which is why I think those links are helpful. Auto-creation is currently linked to.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+section. For information on metric type auto-creation, see `Auto-creation of custom metrics`_.
+
+You must also create a fully parameterized :class:`~gcloud.monitoring.resource.Resource` from one
+of the available ``ResourceDescriptors`` which can be listed using the
```
available resource descriptors? There is no class called ResourceDescriptor**s**

waprin replied:
done

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+You must also create a fully parameterized :class:`~gcloud.monitoring.resource.Resource` from one
+of the available resource descriptors which can be listed using the
+:class:`~gcloud.monitoring.resource.ResourceDescriptor` class, which similarly requires that its
+available labels have values specified. The ``project_id`` label is a special case that should
```
This can be re-written more accurately as:

The label ``project_id`` is ignored when writing and should not be specified.

waprin replied:
done

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+of the available resource descriptors which can be listed using the
+:class:`~gcloud.monitoring.resource.ResourceDescriptor` class, which similarly requires that its
+available labels have values specified. The ``project_id`` label is a special case that should
+not be specified.
```
I am re-writing Ken's suggestion for the paragraph re-write as is - it provides some documentation around picking the appropriate resource type which is useful for a first time user:

You will also need to provide a ``Resource`` instance specifying a monitored resource type as well as values for all of the monitored resource labels, except for ``project_id``, which is ignored when writing. A good choice is to use the underlying physical resource where your application code runs – e.g., a monitored resource type of ``gce_instance`` or ``aws_ec2_instance``. In some limited circumstances, such as when only a single process writes to the custom metric, you may choose to use the global monitored resource type.

waprin replied:
Sounds good but we should link in some docs on where to find the resource descriptors no? I am realizing the class is not as helpful as just the API docs so maybe just link to https://cloud.google.com/monitoring/api/resources ?

supriyagarg replied:
Linking the phrase "monitored resource type" to the API docs link would work.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+
+To write a data point to a custom metric, you must provide an instance of
+:class:`~gcloud.monitoring.metric.Metric` created with a metric type as well as values for
+the labels associated with that metric type. The metric type can either
```
One creates metric descriptors, not metric types.

waprin replied:
Is it inaccurate to say you create the type by creating a metric descriptor? Auto-creation docs read similar to that: 

https://cloud.google.com/monitoring/custom-metrics/creating-metrics#auto-creation

"If you write metric data to a custom metric type that does not yet exist, then a new custom metric type is created for you automatically"

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+To write a data point to a custom metric, you must provide an instance of
+:class:`~gcloud.monitoring.metric.Metric` created with a metric type as well as values for
+the labels associated with that metric type. The metric type can either
+be created using the :class:`~gcloud.monitoring.metric.MetricDescriptor` class or be
```
I'll try to explain why I suggested phrasing this the way I did.

It's confusing to read about creating metric descriptors in text about writing custom metric data. They don't really go together. You may have many application processes writing data for the same metric type, and even more over time, considering restarts. You would only create the metric descriptor once. Especially since they are a limited resource (you only get 500), you might prefer to manage them rather explicitly, in admin tools separate from the application code.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,97 @@ follows::
+monitored resource type as well as values for all of the monitored resource labels, except for
+project_id, which is ignored when writing. A good choice is to use the underlying physical
+resource where your application code runs – e.g., a monitored resource type of gce_instance or
+aws_ec2_instance. In some limited circumstances, such as when only a single process writes to the
```
Please back quotes around project_id, gce_instance and aws_ec2_instance.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+
+You must also create a fully parameterized :class:`~gcloud.monitoring.resource.Resource` from one
+of the available resource descriptors which can be listed using the
+:class:`~gcloud.monitoring.resource.ResourceDescriptor` class, which similarly requires that its
```
- You don't create a `Resource` instance "from" a monitored resource descriptor; you just create one. The monitored resource descriptors are just machine-readable documentation, of a sort.
- You list monitored resource descriptors with the `Client.list_resource_descriptors` method, not "using the `ResourceDescriptor` class".
- In any case, the list of built-in monitored resource descriptors is not very useful in this context, because only a few monitored resource types make sense as a sender of custom metrics. Hence the text I suggested earlier.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+
+For *CUMULATIVE* metrics, the same start time should be re-used repeatedly as more points are
+written to the time series. In this context, the start time is also called the *reset*
+time. Once the cumulative metric is ready to be reset, the *reset* time should be reset by
```
I don't understand what it would mean for a cumulative metric to be "ready to be reset".

This paragraph and the previous one need to be combined and tightened up.

waprin replied:
Trying to communicate when you would set a new start time. I took another shot at it.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,94 @@ follows::
+To write a data point to a custom metric, you must provide an instance of
+:class:`~gcloud.monitoring.metric.Metric` specifying the metric type as well as the values for
+the metric labels. You will need to have either created the metric descriptor earlier ( see the
+`Metric Descriptors`_ section ) or rely on metric type auto-creation, see `Auto-creation of
```
There's extra whitespace inside the parentheses.

The comma seems awfully informal and loose.

waprin replied:
whitespace is fixed

Agreed on the comma, I replaced it with another set of parentheses for consistency with the first part of the sentence. The parentheses I'm also iffy about, we could maybe just split it into separate sentences.

waprin replied:
Also I had  already replaced the comma, commits coming in so make sure you refresh to see latest.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+`project_id`, which is ignored when writing. A good choice is to use the underlying physical
+resource where your application code runs – e.g., a monitored resource type of `gce_instance` or
+`aws_ec2_instance`. In some limited circumstances, such as when only a single process writes to the
+custom metric, you may choose to use the global monitored resource type.
```
`gce_instance`, `aws_ec2_instance`, and most importantly, `global` are code.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+custom metric, you may choose to use the global monitored resource type.
+
+See `Monitored resource types`_ for a list of all monitored resource types available in
+Stackdriver Monitoring.
```
This is a little misleading, because they are _not_ available for writing custom metrics.

waprin replied:
Which ones are available and which ones aren't?

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+reset would otherwise show up as a huge negative spike. For *CUMULATIVE* metrics, the same start
+time should be re-used repeatedly as more points are written to the time series. In this context,
+the start time is also called the *reset* time. Once the metric being measure has its value
+reset, you can specify a new `start_time` to reset the *reset* value. In the examples below, the
```
This text about how you can "reset the `reset` value" doesn't really make sense.

waprin replied:
Yeah I just killed it. I think what you're getting at is the common case is the application code gets the reset time once at startup time, and there is no need to reset it again because it only happens on process startup. If that's the case I guess killing the whole sentence is clearest.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+To write a data point to a custom metric, you must provide an instance of
+:class:`~gcloud.monitoring.metric.Metric` specifying the metric type as well as the values for
+the metric labels. You will need to have either created the metric descriptor earlier (see the
+`Metric Descriptors`_ section ) or rely on metric type auto-creation (see `Auto-creation of
```
Remove space before parenthesis.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+`project_id`, which is ignored when writing. A good choice is to use the underlying physical
+resource where your application code runs – e.g., a monitored resource type of `gce_instance` or
+`aws_ec2_instance`. In some limited circumstances, such as when only a single process writes to the
+custom metric, you may choose to use the `global` monitored resource type.
```
You want double-backquotes, right?

waprin replied:
Yes I was confused when to use which but think I understand now (double backticks for code, single backticks for reference links)

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,92 @@ follows::
+can be used to write :class:`~gcloud.monitoring.timeseries.Point` values.
+
+When writing points, the Python type of the value must match the *value_type* specified
+in the type of the associated `Metric Descriptors`_. For example, a Python float will map to
```
The value type is not specified in the type.

waprin replied:
Re-worded.

waprin commented:
Had to bounce earlier, sorry. Addressed comments again, I am a bit confused about which monitored resources can be used with custom metrics, I don't see anywhere on the doc suggesting that not all of them can be used. Docs actually just link to the same list I link to.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+
+Please refer to the `Metrics`_ documentation for more information.
+
+With a ``Metric`` and ``Resource`` specified, the :class:`~gcloud.monitoring.client.Client`
```
specified -> in hand

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+With a ``Metric`` and ``Resource`` specified, the :class:`~gcloud.monitoring.client.Client`
+can be used to write :class:`~gcloud.monitoring.timeseries.Point` values.
+
+When writing points, the Python type of the value must match the *value_type* of the metric
```
_value_type_ -> _value type_

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+
+``CUMULATIVE`` metrics enable the monitoring system to compute rates of increase on metrics that
+sometimes reset, such as after a process restart. Without cumulative metrics, this
+reset would otherwise show up as a huge negative spike. For ``CUMULATIVE`` metrics, the same start
```
`CUMULATIVE` -> cumulative

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,91 @@ follows::
+In the examples below, the ``end_time`` again defaults to the current time::
+
+    >>> RESET = datetime.utcnow()
+    >>> # Use default ``end_time`` below
```
- Sounds too much like a TODO.
- The double-backticks don't belong in a code sample.

`# The end time defaults to the current time.`

rimey replied:
But the text already says that, so I suggest simply removing the comment.

waprin replied:
Yeah removed, doh on the backticks

rimey commented:
It's really true that custom metrics can be written to only a few of the built-in monitored resource types. Especially because the details are subject to change, this should be addressed in the API documentation, and not here.

waprin commented:
new comments addressed. That does seem like important info to work into the docs.

waprin commented:
@rimey @supriyagarg uploaded implementation of `write_time_series` and `write_point`. Notes:

* I have a better hang of Sphinx now and I tried to be a little more thorough/consistent in doc strings, hopefully less mistakes there
* Maybe could use more unit tests, I wanted a first review pass, and some unit tests are already starting to feel a little redundant so I was hesitant to clutter them up too much with tests that don't accomplish much.
* I felt we needed at least one system test but it was challenging to reduce the flakiness of one. I  relied on tseavers very helpful retry decorators, but I had to increase the number of retries so it will wait a maximum of 64 seconds, though it usually succeeds after waiting 32 seconds. This is a long time for a system test and not ideal, but it seems good to have at least one system test, and if/when the retry logic is right, it should always pass. Unfortunately probably don't want to add too many more, especially not until we resolve #2200 .

I am leaving for 2 weeks starting Monday, and will be 100% off the grid next week, on the grid but not planning to work non-urgent stuff the week after. In the extremely likely case this doesn't get merged today, I'll either do it first thing when I get back or someone can finish it for me.

supriyagarg commented on gcloud/monitoring/client.py:
```diff
@@ -490,3 +490,83 @@ def list_groups(self):
+        """
+        path = '/projects/{project}/timeSeries/'.format(
+            project=self.project)
+        timeseries_dict = [t._to_dict() for t in timeseries_list]
```
t -> timeseries

waprin replied:
done

supriyagarg commented on gcloud/monitoring/timeseries.py:
```diff
@@ -155,3 +199,27 @@ def _from_dict(cls, info):
+                'endTime': _datetime_to_rfc3339(self.end_time)
+            },
+            'value': {
+                _get_typed_value_from_value(self.value): str(self.value)
```
Why are you converting the value to a string? According to the following link, only strings and integers are expected to be strings.

https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TypedValue

waprin replied:
done

supriyagarg commented on gcloud/monitoring/client.py:
```diff
@@ -490,3 +490,83 @@ def list_groups(self):
+
+        Example::
+
+            >>> client.write_time_series([ts1, ts2])
```
You are talking about writing a single time series object & write_point above, but use write_time_series in the code sample. Please make them consistent.

supriyagarg commented on gcloud/monitoring/client.py:
```diff
@@ -490,3 +490,83 @@ def list_groups(self):
+        factory method to create time series objects and the
+        :meth:`~gcloud.monitoring.client.Client.write_time_series` method to
+        write
+        the objects.
```
"the objects." can be moved to the previous line.

waprin replied:
done

waprin commented:
@supriyagarg not sure if you deleted a comment or it got lost in shuffle but I also rearranged the `write_time_series` docstring so that the suggestion to consider `write_point` comes after the example.

supriyagarg commented:
Thanks, that reads better.

waprin commented:
@rimey now turn datetimes into strings in the timezone factories and only pass strings to points.

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,90 @@ follows::
+resource type.
+
+See `Monitored resource types`_ for a list of all monitored resource types available in
+Stackdriver Monitoring.
```
Instead: "...for more information about particular monitored resource types."

This is to avoid the implication that the monitored resource types are available for writing custom metrics.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,90 @@ follows::
+...      'status': 'successful'
+... })
+
+Please refer to the `Metrics`_ documentation for more information.
```
This paragraph interrupts the flow and can be removed.

waprin replied:
done

rimey commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,90 @@ follows::
+descriptor associated with the metric. For example, a Python float will map to ``ValueType.DOUBLE``.
+
+Stackdriver Monitoring supports several *metric kinds*: ``GAUGE``, ``CUMULATIVE``, and ``DELTA``.
+However, ``DELTA`` custom metrics are not supported.
```
"...is not supported for custom metrics."

waprin replied:
done

rimey commented on gcloud/monitoring/client.py:
```diff
@@ -490,3 +496,84 @@ def list_groups(self):
+        instead.
+
+        :type timeseries_list:
+            list of :class:`~gcloud.monitoring.timeseries.TimeSeries`.
```
We're omitting periods after type descriptions.

waprin replied:
done

rimey commented on gcloud/monitoring/metric.py:
```diff
@@ -347,3 +347,16 @@ def _from_dict(cls, info):
+            'type': self.type,
+            'labels': self.labels
+        }
+        return info
```
```
        return {
            'type': self.type,
            'labels': self.labels,
        }
```

Note the comma after the last item.

waprin replied:
done

rimey commented on gcloud/monitoring/resource.py:
```diff
@@ -186,3 +186,16 @@ def _from_dict(cls, info):
+            'type': self.type,
+            'labels': self.labels
+        }
+        return info
```
```
        return {
            'type': self.type,
            'labels': self.labels,
        }
```

Note the comma after the last item.

waprin replied:
done

rimey commented on gcloud/monitoring/client.py:
```diff
@@ -490,3 +496,84 @@ def list_groups(self):
         return Group._list(self)
+
+    def write_time_series(self, timeseries_list):
+        """Writes a list of time series objects to the API.
```
Writes -> Write

waprin replied:
done

rimey commented on gcloud/monitoring/client.py:
```diff
@@ -490,3 +496,84 @@ def list_groups(self):
+    def write_point(self, metric, resource, value,
+                    end_time=None,
+                    start_time=None):
+        """Writes a single point for a metric to the API.
```
Writes -> Write

waprin replied:
done

rimey commented on gcloud/monitoring/timeseries.py:
```diff
@@ -89,6 +89,23 @@ def header(self, points=None):
+            'points': [point._to_dict() for point in self.points]
+        }
+
+        return info
```
```
        return {
            'metric': self.metric._to_dict(),
            'resource': self.resource._to_dict(),
            'points': [point._to_dict() for point in self.points],
        }
```

waprin replied:
done

rimey commented on gcloud/monitoring/timeseries.py:
```diff
@@ -123,6 +140,32 @@ def __repr__(self):
 
 
+def _get_typed_value_from_value(value):
+    """Infers the typed value for a point from the Python type.
```
Infers -> Infer

waprin replied:
rewrote

rimey commented on gcloud/monitoring/timeseries.py:
```diff
@@ -123,6 +140,32 @@ def __repr__(self):
+   :rtype: str
+   :returns: The string that signifies the appropriate typed value.
+    """
+    typed_value_map = {
```
Maybe make this a module-level constant.

waprin replied:
Leaning against it due to a preference for keeping scope of variables as narrow as possible, however if you feel strongly I can still change.

rimey commented on gcloud/monitoring/timeseries.py:
```diff
@@ -123,6 +140,32 @@ def __repr__(self):
         )
 
 
+def _get_typed_value_from_value(value):
```
This doesn't do what its name suggests. A better name might be `_get_type_key`.

rimey replied:
I wonder. Would you perhaps prefer to have a function that actually returns the typed value? For example, given `3.14`, it would return `{'doubleValue': 3.14}`.

waprin replied:
Agreed, switched to that

rimey commented on gcloud/monitoring/timeseries.py:
```diff
@@ -155,3 +198,31 @@ def _from_dict(cls, info):
+        :rtype: dict
+        :returns: The dictionary representation of the point object.
+        """
+        typed_value = _get_typed_value_from_value(self.value)
```
A better name for this variable might be `type_key`.

rimey commented on system_tests/monitoring.py:
```diff
@@ -181,6 +183,53 @@ def test_create_and_delete_metric_descriptor(self):
+        retry_404(descriptor.delete)()
+
+        with self.assertRaises(NotFound):
+            descriptor.delete()
```
I'd prefer to have any changes to `test_create_and_delete_metric_descriptor` in a separate pull request.

waprin replied:
Shouldn't have even been in here

rimey commented on system_tests/monitoring.py:
```diff
@@ -181,6 +183,53 @@ def test_create_and_delete_metric_descriptor(self):
+        )
+
+        descriptor.create()
+        retry_404(descriptor._fetch)(client, METRIC_TYPE)
```
I think it would make more sense to retry `write_point`.

waprin replied:
write_point never failed for me, most of the retry logic in here was based on pratical desire to get the test working

rimey replied:
I've been digging deeper into this. Let's be clear that this statement should not be necessary! Nevertheless, the test failed without it.

Strictly speaking, it's not actually the right thing to check. Just because `GetMetricDescriptor` succeeds, that doesn't mean that another call to `GetMetricDescriptor` would succeed, or that a call to `CreateTimeSeries` will succeed.

If you feel more comfortable leaving this as is, that's fine, but my suggestion would be to replace it with `time.sleep(5)`.

waprin replied:
Ah sorry I did miss this comment but I'm confused why you would prefer to replace it with time.sleep(). I think the retry decorators are a little better since they can use less time sometimes and are less flaky with the exponential backoff. 

I think if anything you're write I should get rid of this then retry write_point? That makes sense to me

rimey replied:
I feel time.sleep(5) is clearest. GetMetricDescriptors may return quickly, after which CreateTimeSeries can still fail.

dhermes replied:
@rimey :-1: That's the whole purpose of the `retry` helpers. If there is a missing feature, let us know.

rimey replied:
I'm not following your logic. Why do you think GetMetricDescriptors returning a value is a suitable prerequisite to calling CreateTimeSeries?

dhermes replied:
@rimey Sorry for not being clearer, :-1: to using `time.sleep(5)`. No comment on any of the rest of your preferences.

rimey replied:
@dhermes What alternative would you propose?

dhermes replied:
Using the `retry` helpers

rimey replied:
Retrying what?

There's no point to calling GetMonitoredResource here. It tells you nothing about whether the time series datastore knows about the custom metric.

The underlying big picture here is that you normally only create a descriptor for a monitored resource once. In production, you wouldn't care if the first call or two to CreateTimeSeries fails. Experiments, code samples, and system tests are different from production code in that respect.

dhermes replied:
:+1: In the GitHub UI I see this comment directly below `retry_404(descriptor._fetch)(client, METRIC_TYPE)`. Maybe this is out-of-date?

Carry on as if I never dropped by

waprin replied:
Ok I just killed it, seems the simplest thing to do since it's not really necessary.

rimey commented on system_tests/monitoring.py:
```diff
@@ -181,6 +183,53 @@ def test_create_and_delete_metric_descriptor(self):
+                return len(list(result)) > 0
+            retry_result = RetryResult(_has_timeseries, max_tries=7)(
+                client.query)
+            return RetryErrors(BadRequest)(retry_result)
```
You're seeing 400s?!?

waprin replied:
Yes, metric descriptor doesn't exist yet

waprin replied:
400 The provided filter doesn't refer to any known metric. (

rimey replied:
Does the advice above about `time.sleep(5)` resolve this?

rimey commented on system_tests/monitoring.py:
```diff
@@ -181,6 +183,53 @@ def test_create_and_delete_metric_descriptor(self):
+        self.assertEqual(len(timeseries_list), 1)
+        timeseries = timeseries_list[0]
+        self.assertEqual(timeseries.metric, metric)
+        # resource labels will not be equal.
```
Instead of saying this, test that they are equal except for the additional project ID in the query result.

waprin replied:
done

rimey commented on system_tests/monitoring.py:
```diff
@@ -181,6 +183,53 @@ def test_create_and_delete_metric_descriptor(self):
+        self.assertEqual(timeseries.resource.type, resource.type)
+        self.assertEqual(timeseries.points[0].value, VALUE)
+
+        retry_404(descriptor.delete)()
```
You believe that `delete` can return a 404 at this point? I doubt it can.

waprin replied:
I don't think so either but `test_create_and_delete_metric_descriptor` retries the delete so I was just following that lead.

rimey commented:
I made another pass. PTAL.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,90 @@ follows::
+``GAUGE`` metrics represent only a single point in time, so only the ``end_time`` should be
+specified::
+
+    >>> client.write_point(metric=metric, resource=resource, 3.14, end_time=end) # API call
```
`end` -> `end_time`

supriyagarg replied:
Also, add value=3.14 since otherwise this code will raise errors.

waprin replied:
done

waprin commented:
@rimey @supriyagarg im back! Thanks for patience, addressed review comments, take a look when you get a chance.

waprin commented:
Also had to rebase on the rename so might be worth keeping in mind anywhere I missed that "ctrl-f" for gcloud is empty though so we should be good.

waprin commented on system_tests/monitoring.py:
```diff
@@ -176,6 +178,48 @@ def test_create_and_delete_metric_descriptor(self):
+        del timeseries.resource.labels['project_id']
+        self.assertEqual(timeseries.resource, resource)
+
+        descriptor.delete()
```
not sure where a lot of my review comments went, but I took away the retry here, but the above test does a retry on descriptor.delete() so I thought I should copy that, but not sure.

supriyagarg commented on docs/monitoring-usage.rst:
```diff
@@ -278,3 +278,88 @@ follows::
+specified::
+
+    >>> client.write_point(metric=metric, resource=resource,
+    ... value=3.14, end_time=end_time) # API call
```
Please indent this so that `value=` lines up under `metric=`.

waprin replied:
done

rimey commented on google/cloud/monitoring/metric.py:
```diff
@@ -348,3 +348,16 @@ def _from_dict(cls, info):
+            'type': self.type,
+            'labels': self.labels,
+        }
+        return info
```
```
        return {
            'type': self.type,
            'labels': self.labels,
        }
```

waprin replied:
done

rimey commented on google/cloud/monitoring/resource.py:
```diff
@@ -187,3 +187,16 @@ def _from_dict(cls, info):
+            'type': self.type,
+            'labels': self.labels,
+        }
+        return info
```
```
        return {
            'type': self.type,
            'labels': self.labels,
        }
```

waprin replied:
done

rimey commented on google/cloud/monitoring/timeseries.py:
```diff
@@ -124,6 +141,40 @@ def __repr__(self):
 
 
+def _make_typed_value(value):
+    """Creates a dict representing a TypedValue API object from a value.
```
Creates → Create

rimey replied:
Maybe drop "from a value".

waprin replied:
done

rimey commented on google/cloud/monitoring/timeseries.py:
```diff
@@ -124,6 +141,40 @@ def __repr__(self):
+        value = str(value)
+    return {
+        type_: value
+    }
```
```
    return {type_: value}
```

waprin replied:
done

waprin commented:
@rimey missed those notifications, addressed those comments. Bit confused if I'm doing something wrong with new Github interface to note I addressed your comments.

rimey commented on google/cloud/monitoring/timeseries.py:
```diff
@@ -124,6 +141,38 @@ def __repr__(self):
+    type_ = typed_value_map[type(value)]
+    if type_ == "int64Value":
+        value = str(value)
+    return { type_: value }
```
You have extra whitespace here and this is failing lint.

waprin commented:
@rimey as mentioned offline just removed the descriptor fetch, seems to be fine, added retry_500 to write_point which happened about once out of ten times.

dhermes commented:
@rimey RE: doing a squash merge, that subverts our typical way of gathering all relevant changes for a release. We can work around it, but it's easiest if all PRs are just done via a merge rather than a squash.

rimey commented:
@dhermes Are you saying we should prefer an old-fashioned approach where the author rebases to avoid having lots of small commits – e.g., as a result of responding to review comments?

dhermes commented:
Yes, though we expect the typical author to be experienced enough to hanlde this and break PRs into logical pieces. Not sure what makes it old or new fashioned.
